File: packages/sdk-go/examples/webmcp.go
package main

import (
	"context"
	"errors"
	"fmt"
	"log"

	stagehand "github.com/browserbase/stagehand/packages/sdk-go/v4"
)

const webMCPTestSite = "https://browserbase.github.io/stagehand-eval-sites/sites/webmcp-test/"

func main() {
	if err := run(context.Background()); err != nil {
		log.Fatal(err)
	}
}

func run(ctx context.Context) (err error) {
	browser, err := stagehand.LaunchLocalBrowser(ctx, &stagehand.LocalBrowserLaunchOptions{Headless: false})
	if err != nil {
		return err
	}
	defer func() { err = errors.Join(err, browser.Close(ctx)) }()

	client, err := stagehand.Create(ctx, stagehand.CreateOptions{Browser: browser})
	if err != nil {
		return err
	}
	defer func() { err = errors.Join(err, client.Close(ctx)) }()

	browserContext, err := browser.Context()
	if err != nil {
		return err
	}
	pages, err := browserContext.Pages(ctx)
	if err != nil {
		return err
	}
	if len(pages) == 0 {
		return errors.New("Stagehand initialized without an active page")
	}
	page := pages[0]
	if _, err := page.Goto(ctx, webMCPTestSite, nil); err != nil {
		return err
	}

	tools, err := page.Tools(ctx, &stagehand.WebMCPToolsOptions{Timeout: 5_000})
	if err != nil {
		return err
	}
	var calculateSum *stagehand.WebMCPTool
	for _, tool := range tools {
		if tool.Descriptor().Name == "calculateSum" {
			calculateSum = tool
			break
		}
	}
	if calculateSum == nil {
		return errors.New("calculateSum was not registered by the page")
	}

	invocation, err := calculateSum.Invoke(ctx, stagehand.WebMCPInput{"a": 19, "b": 23})
	if err != nil {
		return err
	}
	result, err := invocation.Result(ctx, nil)
	if err != nil {
		return err
	}

	fmt.Printf("status: %s\noutput: %s\n", result.Status, result.Output)
	return nil
}

File: packages/sdk-ruby/examples/webmcp.rb
# frozen_string_literal: true

# WebMCP: pages can register tools (navigator.modelContext); Stagehand lists
# them via page.tools and invokes them. Ruby port of
# packages/sdk-python/examples/webmcp.py — no LLM needed.
#
#   ruby examples/webmcp.rb                # local Chrome (headed, like the sibling SDKs)
#   ruby examples/webmcp.rb --browserbase  # Browserbase session

require "json"

require_relative "../lib/stagehand"
require_relative "example_helpers"

WEBMCP_TEST_SITE = "https://browserbase.github.io/stagehand-eval-sites/sites/webmcp-test/"

browser =
  if ARGV.include?("--browserbase")
    api_key = ExampleHelpers.env("BROWSERBASE_API_KEY") or abort "Set BROWSERBASE_API_KEY."
    puts "Creating a Browserbase session..."
    Stagehand::Browserbase.launch(api_key: api_key)
  else
    puts "Launching local Chrome..."
    Stagehand::LocalBrowser.launch(headless: false)
  end

begin
  stagehand = Stagehand.create(browser: browser, log_level: ENV.fetch("STAGEHAND_LOG_LEVEL", "warn"))
  begin
    page = browser.context.pages.first
    raise "Stagehand initialized without an active page" if page.nil?
    page.goto(WEBMCP_TEST_SITE)

    tools = page.tools(timeout: 5_000)
    puts "tools -> #{tools.map(&:name).join(", ")}"
    calculate_sum = tools.find { |tool| tool.name == "calculateSum" }
    raise "calculateSum was not registered by the page" if calculate_sum.nil?

    invocation = calculate_sum.invoke(input: { a: 19, b: 23 })
    result = invocation.result

    puts JSON.pretty_generate(result.to_wire)
  ensure
    stagehand.close
  end
ensure
  browser.close
end
puts "Closed."
